src/main/java/com/redhat/devtools/lsp4ij/operations/inlayhint/LSPInlayHintsProvider.java

```diff
@@ -41,63 +42,130 @@
+    private static final Key<CancellationSupport> CANCELLATION_SUPPORT_KEY = Key.create(LSPInlayHintsProvider.class.getName() + "-CANCELLATION_SUPPORT_KEY");
+
+    private static final Key<CachedResolvedInlayHints> CACHED_RESOLVED_INLAY_HINTS_KEY = Key.create(LSPInlayHintsProvider.class.getName() + "-CACHED_RESOLVED_INLAY_HINTS_KEY");
+    private static final InlayHint RESOLVE_INLAYHINT_NOT_READY = new InlayHint();
 
     public LSPInlayHintsProvider() {
         super(CANCELLATION_SUPPORT_KEY);
     }
 
 
     @Override
-    protected void doCollect(@NotNull VirtualFile file, @NotNull Project project, @NotNull Editor editor, @NotNull PresentationFactory factory, @NotNull InlayHintsSink inlayHintsSink, @NotNull CancellationSupport cancellationSupport) throws InterruptedException {
-        Document document = editor.getDocument();
-        URI fileUri = LSPIJUtils.toUri(file);
-        Range viewPortRange = getViewPortRange(editor);
-        InlayHintParams param = new InlayHintParams(LSPIJUtils.toTextDocumentIdentifier(file), viewPortRange);
-        BlockingDeque<Pair<InlayHint, LanguageServer>> pairs = new LinkedBlockingDeque<>();
+    protected void doCollect(@NotNull VirtualFile file,
+                             @NotNull PsiFile psiFile,
+                             @NotNull Project project,
+                             @NotNull Editor editor,
+                             @NotNull PresentationFactory factory,
+                             @NotNull InlayHintsSink inlayHintsSink,
+                             @NotNull CancellationSupport cancellationSupport) throws InterruptedException {
+
+        // Get from cache inlay hints or create them
+        final List<Pair<Integer, InlayHintData>> inlayHints = getOrCreateInlayHints(file, psiFile, project, editor, cancellationSupport);
 
-        CompletableFuture<Void> future = collect(project, file, param, pairs, cancellationSupport);
-        List<Pair<Integer, Pair<InlayHint, LanguageServer>>> inlayHints = createInlayHints(document, pairs, future);
+        // Render code lenses and collect all codeLens/resolve futures which are not finished
+        List<CompletableFuture> resolvedInlayHintFutures = new ArrayList<>();
         inlayHints.stream()
                 .collect(Collectors.groupingBy(p -> p.first))
                 .forEach((offset, list) ->
-                        inlayHintsSink.addInlineElement(offset, false, toPresentation(editor, list, factory, cancellationSupport), false));
+                        inlayHintsSink.addInlineElement(offset,
+                                false,
+                                toPresentation(editor, list, factory, cancellationSupport),
+                                false));
+
+        if (!resolvedInlayHintFutures.isEmpty() && !cancellationSupport.isCanceled()) {
+            // There are some inlayHint/resolve which are not finished, wait for all finish of those futures and refresh the inlay hints
```

Comment:
Wait for all unfinished inlayHint/resolve futures to complete and refresh the inlay hints